File: packages/hook/src/lib/check.ts
/**
 * Shared check-path helpers.
 *
 * Both `exec check` and `task check` follow the same state machine:
 *   1. Stop-event recursion guard → skip if already in a stop continuation
 *   2. Consume sentinel → missing / pending / pass / fail
 */

import { readMarker } from "../commands/scope";
import type { AgentEvent, HookAdapter } from "./adapter";
import type { ResolvedConfig, ResolvedExec } from "./config";
import { getTriggerPatterns } from "./config";
import { computeFingerprint, detectChanges } from "./git";
import { log } from "./log";
import type { SentinelData } from "./sentinel";
import { incrementBlockCount, readSentinel, resetBlockCount } from "./sentinel";

// ---------------------------------------------------------------------------
// Sentinel evaluation result
// ---------------------------------------------------------------------------

/** Discriminated union describing the outcome of evaluating a sentinel. */
export type SentinelCheckResult =
	| { kind: "missing" }
	| { kind: "pending" }
	| { kind: "pass" }
	| { kind: "fail"; sentinel: SentinelData };

/**
 * Evaluate a consumed sentinel and return a structured result.
 *
 * Staleness checks (applied in order):
 *
 * 1. **Session-aware:** When `currentSessionId` is provided, the sentinel must
 *    carry a matching `sessionId` — otherwise it is treated as stale (`"missing"`).
 *
 * 2. **Content-aware:** When `currentContentHash` is provided, sentinels must
 *    also carry a matching `contentHash` — otherwise they are treated as stale.
 *    This prevents bait-and-switch: an agent cannot run tests on clean code,
 *    then modify files and commit with the stale passing sentinel. Sentinels
 *    without a `contentHash` are also rejected.
 */
export function evaluateSentinel(
	sentinel: SentinelData | undefined,
	currentSessionId?: string,
	currentContentHash?: string,
): SentinelCheckResult {
	if (!sentinel) return { kind: "missing" };

	// Session-aware staleness: the sentinel must belong to the current session.
	// Missing or mismatched sessionId → treat as stale.
	if (currentSessionId && (!sentinel.sessionId || sentinel.sessionId !== currentSessionId)) {
		return { kind: "missing" };
	}

	// Pending sentinels are written before the command finishes, so they
	// never carry a contentHash — check them before hash validation.
	if (sentinel.status === "pending") return { kind: "pending" };

	// Content-aware staleness: if the caller provides a content hash,
	// the sentinel must also have one and it must match. A sentinel
	// without a contentHash is treated as stale — this closes the
	// loophole where a hand-crafted sentinel without a hash bypasses
	// validation entirely. Only applies to terminal states (pass/fail).
	if (currentContentHash) {
		if (!sentinel.contentHash || sentinel.contentHash !== currentContentHash) {
			return { kind: "missing" };
		}
	}

	if (sentinel.status === "pass") return { kind: "pass" };
	return { kind: "fail", sentinel };
}

// ---------------------------------------------------------------------------
// Exec pre-evaluation pipeline (pure)
// ---------------------------------------------------------------------------

/** Discriminated union describing the outcome of pre-evaluating an exec spec. */
export type ExecCheckVerdict =
	| { kind: "skip-trigger" }
	| { kind: "skip-no-changes" }
	| { kind: "missing" }
	| { kind: "pending"; sentinel?: SentinelData }
	| { kind: "pass" }
	| { kind: "fail"; sentinel: SentinelData };

/** Optional pre-computed values to avoid redundant git operations. */
export type PreEvalCache = {
	/** Pre-computed result of `detectChanges()`. When undefined, detection runs inline. */
	hasChanges?: boolean;
	/** Pre-computed result of `computeFingerprint()`. When undefined, computed inline. */
	contentHash?: string;
};

/**
 * Pure pre-evaluation pipeline for an exec spec.
 *
 * Runs the full check logic — trigger matching, change detection, sentinel
 * reading, session-aware and content-aware staleness — and returns a verdict
 * without any side effects (no `process.exit`, no sentinel removal).
 *
 * Both standalone `exec check` and `sync check` delegate to this function.
 *
 * @param config     Resolved project config.
 * @param adapter    Hook adapter (used only for trigger matching, not for blocking).
 * @param event      The current agent event.
 * @param exec       Resolved exec configuration (command, fileExt, always, etc.).
 * @param flags      Parsed exec flags (name, staged, trigger, etc.).
 * @param cache      Optional pre-computed change detection and fingerprint values.
 */
export async function preEvaluateExec(
	config: ResolvedConfig,
	adapter: HookAdapter,
	event: AgentEvent,
	exec: ResolvedExec,
	flags: { name: string; staged?: boolean; on?: string; trigger?: string },
	cache?: PreEvalCache,
): Promise<ExecCheckVerdict> {
	// Trigger matching — skip if the event doesn't match the trigger filter.
	const triggerPatterns = resolveTriggerPatterns(`exec:${flags.name}`, config, flags);
	if (triggerPatterns.length > 0 && !matchesTrigger(adapter, event, triggerPatterns)) {
		return { kind: "skip-trigger" };
	}

	// Skip-if-no-changes.
	// At push time the working tree is always clean, so detectChanges would
	// short-circuit — skipping sentinel/fingerprint validation entirely. We
	// still keep it for non-push events as a performance optimization.
	const isPush = matchesTrigger(adapter, event, ["git push"]);
	if (!exec.always && !isPush) {
		const hasChanges =
			cache?.hasChanges ??
			(await detectChanges({
				cwd: config.projectDir,
				fileExt: exec.fileExt,
				staged: flags.staged,
			}));
		if (!hasChanges) {
			return { kind: "skip-no-changes" };
		}
	}

	// Read the individual sentinel.
	const sentinel = readSentinel(config.sentinelDir, config.projectDir, flags.name);

	// Session-aware staleness.
	const marker = readMarker(config.projectDir);
	const currentSessionId = marker?.sessionId;

	// Content-aware staleness: compute or use cached fingerprint.
	const contentHash =
		cache?.contentHash ??
		(await computeFingerprint({
			cwd: config.projectDir,
			staged: flags.staged,
			fileExt: exec.fileExt,
		}));

	const result = evaluateSentinel(sentinel, currentSessionId, contentHash);

	switch (result.kind) {
		case "missing":
			return { kind: "missing" };
		case "pending":
			return { kind: "pending", sentinel };
		case "pass":
			return { kind: "pass" };
		case "fail":
			return { kind: "fail", sentinel: result.sentinel };
	}
}

// ---------------------------------------------------------------------------
// Stop-event recursion guard
// ---------------------------------------------------------------------------

/**
 * Guard against infinite recursion on Stop events.
 *
 * When a Stop hook blocks, Claude Code re-fires the Stop event in a
 * "stop continuation" (with `stop_hook_active = true`). Without a guard,
 * blocking again creates an infinite loop.
 *
 * Behavior depends on `limit`:
 *   - `limit > 0` — Let `blockWithLimit` enforce the limit as usual.
 *   - `limit = 0` (unlimited) — Auto-allow immediately to prevent infinite loop.
 */
export function guardStopEvent(
	tag: string,
	adapter: HookAdapter,
	event: AgentEvent,
	limit: number,
): void {
	if (adapter.isStopRecursion(event)) {
		if (limit > 0) {
			log(tag, `stop_hook_active=true, limit=${limit} — deferring to blockWithLimit`);
			return;
		}
		log(tag, "stop_hook_active=true, limit=0 — auto-allowing (infinite-loop prevention)");
		adapter.allow();
	}
}

// ---------------------------------------------------------------------------
// Trigger pattern resolution
// ---------------------------------------------------------------------------

/** Common flags used by trigger resolution. */
type TriggerFlags = { on?: string; trigger?: string };

/**
 * Resolve trigger patterns from `--on` (named group) or `--trigger` (inline).
 * Returns an empty array when no trigger filter is set (matches everything).
 */
export function resolveTriggerPatterns(
	tag: string,
	config: ResolvedConfig,
	flags: TriggerFlags,
): string[] {
	if (flags.trigger) return [flags.trigger];
	if (flags.on) {
		const patterns = getTriggerPatterns(config, flags.on);
		if (!patterns) {
			log(tag, `Warning: trigger group "${flags.on}" not found in config`);
			return [];
		}
		return patterns;
	}
	return [];
}

// ---------------------------------------------------------------------------
// Trigger matching
// ---------------------------------------------------------------------------

/**
 * Check whether the current hook event matches any of the given trigger patterns.
 *
 * Trigger patterns are matched against the shell command string (case-insensitive
 * substring). Returns `true` if any pattern matches, or if `patterns` is empty.
 */
export function matchesTrigger(
	adapter: HookAdapter,
	event: AgentEvent,
	patterns: string[],
): boolean {
	if (patterns.length === 0) return true;

	const command = adapter.isShellToolCall(event) ? adapter.getShellCommand(event) : undefined;
	if (typeof command !== "string") return false;

	const lower = command.toLowerCase();
	return patterns.some((p) => lower.includes(p.toLowerCase()));
}

// ---------------------------------------------------------------------------
// Block-with-limit
// ---------------------------------------------------------------------------

/**
 * Block (or auto-allow if limit exceeded).
 *
 * Increments the consecutive failure counter and checks against the
 * configured limit. When `limit > 0` and the counter exceeds it,
 * auto-allows and resets the counter.
 */
export function blockWithLimit(
	tag: string,
	adapter: HookAdapter,
	config: ResolvedConfig,
	name: string,
	limit: number,
	reason: string,
): never {
	const count = incrementBlockCount(config.sentinelDir, config.projectDir, name);
	if (limit > 0 && count > limit) {
		log(
			tag,
			`WARNING: requirement not satisfied — block count ${count} exceeds limit ${limit}, auto-allowing`,
		);
		resetBlockCount(config.sentinelDir, config.projectDir, name);
		adapter.allow();
	}
	log(tag, `Action: block (${count}/${limit || "∞"}) — agent must re-run`);
	adapter.block(withProjectHeader(config.projectDir, reason));
}

/**
 * Block without incrementing the failure counter.
 * Used for intermediate / transient states (missing, pending).
 */
export function blockNoCount(
	tag: string,
	adapter: HookAdapter,
	reason: string,
	projectDir?: string,
): never {
	log(tag, "Action: block (no counter increment — transient state)");
	adapter.block(projectDir ? withProjectHeader(projectDir, reason) : reason);
}

// ---------------------------------------------------------------------------
// Project header
// ---------------------------------------------------------------------------

/**
 * Prepend a `[project: <dir>]` header to a block reason.
 */
function withProjectHeader(projectDir: string, reason: string): string {
	return `[project: ${projectDir}]\n${reason}`;
}

File: internal/hook/check.go
package hook

import (
	"fmt"
	"os"
	"strings"
)

// CheckResult is the outcome of evaluating a sentinel.
type CheckResult struct {
	Kind     string // "missing", "pending", "pass", "fail"
	Sentinel *SentinelData
}

// evaluateSentinel evaluates a sentinel with session and content staleness checks.
func evaluateSentinel(sentinel *SentinelData, currentSessionID, currentContentHash string) CheckResult {
	if sentinel == nil {
		return CheckResult{Kind: "missing"}
	}

	// Session-aware staleness
	if currentSessionID != "" {
		if sentinel.SessionID == "" || sentinel.SessionID != currentSessionID {
			return CheckResult{Kind: "missing"}
		}
	}

	// Pending sentinels don't carry content hashes
	if sentinel.Status == "pending" {
		return CheckResult{Kind: "pending", Sentinel: sentinel}
	}

	// Content-aware staleness
	if currentContentHash != "" {
		if sentinel.ContentHash == "" || sentinel.ContentHash != currentContentHash {
			return CheckResult{Kind: "missing"}
		}
	}

	if sentinel.Status == "pass" {
		return CheckResult{Kind: "pass", Sentinel: sentinel}
	}
	return CheckResult{Kind: "fail", Sentinel: sentinel}
}

// ExecCheckVerdict is the outcome of pre-evaluating an exec spec.
type ExecCheckVerdict struct {
	Kind     string // "skip-trigger", "skip-no-changes", "missing", "pending", "pass", "fail"
	Sentinel *SentinelData
}

// preEvaluateExec runs the full check pipeline for an exec spec.
func preEvaluateExec(cfg *ResolvedConfig, event map[string]interface{}, exec ExecConfig, name string, staged bool, on, trigger string, hasChangesCache *bool, contentHashCache string) ExecCheckVerdict {
	// Trigger matching
	patterns := resolveTriggerPatterns(cfg, on, trigger)
	if len(patterns) > 0 && !matchesTrigger(event, patterns) {
		return ExecCheckVerdict{Kind: "skip-trigger"}
	}

	// Skip if no changes
	if !exec.Always {
		var hasChanges bool
		if hasChangesCache != nil {
			hasChanges = *hasChangesCache
		} else {
			var err error
			hasChanges, err = detectChanges(cfg.ProjectDir, exec.FileExt, staged)
			if err != nil {
				hasChanges = false
			}
		}
		if !hasChanges {
			return ExecCheckVerdict{Kind: "skip-no-changes"}
		}
	}

	// Read sentinel
	sentinel := readSentinel(cfg.SentinelDir, cfg.ProjectDir, name)

	// Session-aware staleness
	marker := ReadMarker(cfg.ProjectDir)
	currentSessionID := ""
	if marker != nil {
		currentSessionID = marker.SessionID
	}

	// Content-aware staleness
	contentHash := contentHashCache
	if contentHash == "" {
		contentHash = computeFingerprint(cfg.ProjectDir, staged, exec.FileExt)
	}

	result := evaluateSentinel(sentinel, currentSessionID, contentHash)

	// Command validation: prevent --cmd bypass
	if (result.Kind == verdictPass || result.Kind == verdictFail) && sentinel != nil {
		sentinelCmd := sentinel.ConfiguredCommand
		if sentinelCmd == "" {
			sentinelCmd = sentinel.Command
		}
		if sentinelCmd != "" && sentinelCmd != exec.Command {
			return ExecCheckVerdict{Kind: "missing"}
		}
	}

	return ExecCheckVerdict(result)
}

// resolveTriggerPatterns resolves trigger patterns from --on or --trigger flags.
func resolveTriggerPatterns(cfg *ResolvedConfig, on, trigger string) []string {
	if trigger != "" {
		return []string{trigger}
	}
	if on != "" {
		patterns, ok := cfg.Triggers[on]
		if !ok {
			return nil
		}
		return patterns
	}
	return nil
}

// matchesTrigger checks whether the event matches any trigger patterns.
// Patterns are matched as case-insensitive substrings against the shell command.
func matchesTrigger(event map[string]interface{}, patterns []string) bool {
	if len(patterns) == 0 {
		return true
	}

	command := extractShellCommand(event)
	if command == "" {
		return false
	}

	lower := strings.ToLower(command)
	for _, p := range patterns {
		if strings.Contains(lower, strings.ToLower(p)) {
			return true
		}
	}
	return false
}

// extractShellCommand extracts the shell command from a tool_input payload.
func extractShellCommand(event map[string]interface{}) string {
	toolInput, ok := event["tool_input"].(map[string]interface{})
	if !ok {
		return ""
	}
	cmd, _ := toolInput["command"].(string)
	return cmd
}

// Response represents the action to take: allow or block.
type Response struct {
	Action  string // "allow" or "block"
	Message string // block reason (only for "block")
}

// blockWithLimit blocks or auto-allows if limit exceeded.
// Returns the response to emit.
func blockWithLimit(cfg *ResolvedConfig, name string, limit int, reason string) Response {
	count := incrementBlockCount(cfg.SentinelDir, cfg.ProjectDir, name)
	if limit > 0 && count > limit {
		resetBlockCount(cfg.SentinelDir, cfg.ProjectDir, name)
		return Response{Action: "allow"}
	}
	return Response{
		Action:  "block",
		Message: withProjectHeader(cfg.ProjectDir, reason),
	}
}

// blockNoCount blocks without incrementing the counter.
func blockNoCount(projectDir, reason string) Response {
	return Response{
		Action:  "block",
		Message: withProjectHeader(projectDir, reason),
	}
}

func withProjectHeader(projectDir, reason string) string {
	return fmt.Sprintf("[project: %s]\n%s", projectDir, reason)
}

// guardStopEvent guards against infinite recursion on Stop events.
// Returns an allow response if recursion should be stopped, nil otherwise.
func guardStopEvent(event map[string]interface{}, limit int) *Response {
	if !isStopRecursion(event) {
		return nil
	}
	if limit > 0 {
		// Let blockWithLimit handle it
		return nil
	}
	resp := Response{Action: "allow"}
	return &resp
}

func isStopRecursion(event map[string]interface{}) bool {
	// Claude Code sets stop_hook_active=true when re-firing Stop after a block
	active, _ := event["stop_hook_active"].(bool)
	return active
}

// emitResponse writes the hook response to stderr (for block) and exits.
func emitResponse(resp Response) error {
	if resp.Action == "allow" {
		return nil // exit 0
	}
	// Write block reason to stderr and return an error that signals exit 2
	fmt.Fprint(os.Stderr, resp.Message)
	return &BlockError{Message: resp.Message}
}

// BlockError signals that the hook should exit with code 2.
type BlockError struct {
	Message string
}

func (e *BlockError) Error() string {
	return e.Message
}

// getBaselineFingerprint retrieves the baseline fingerprint from state.
func getBaselineFingerprint(sentinelDir, projectDir, eventName string) string {
	state := readState(sentinelDir, projectDir)
	eventData, ok := state[eventName].(map[string]interface{})
	if !ok {
		return ""
	}
	entries, ok := eventData["__entries"].([]interface{})
	if !ok || len(entries) == 0 {
		return ""
	}
	last, ok := entries[len(entries)-1].(map[string]interface{})
	if !ok {
		return ""
	}
	fp, _ := last["fingerprint"].(string)
	return fp
}
